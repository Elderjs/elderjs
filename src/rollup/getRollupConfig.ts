import svelte from 'rollup-plugin-svelte';
import { nodeResolve } from '@rollup/plugin-node-resolve';
import commonjs from '@rollup/plugin-commonjs';
import { terser } from 'rollup-plugin-terser';
import babel from 'rollup-plugin-babel';
import multiInput from 'rollup-plugin-multi-input';
import replace from '@rollup/plugin-replace';
import json from '@rollup/plugin-json';
import glob from 'glob';
import path from 'path';
import fs from 'fs-extra';
import del from 'del';
import defaultsDeep from 'lodash.defaultsdeep';
import { getElderConfig } from '../index';
import { getDefaultRollup } from '../utils/validations';
import getPluginLocations from '../utils/getPluginLocations';
import elderSvelte from './rollupPlugin';

const production = process.env.NODE_ENV === 'production' || !process.env.ROLLUP_WATCH;
const elderJsDir = path.resolve(process.cwd(), './node_modules/@elderjs/elderjs/');

const babelIE11 = babel({
  extensions: ['.js', '.mjs', '.html', '.svelte'],
  runtimeHelpers: true,
  exclude: ['node_modules/@babel/**', 'node_modules/core-js/**', /\/core-js\//],
  presets: [
    [
      '@babel/preset-env',
      {
        targets: {
          browsers: ['> 0.25%', 'not dead', 'IE 11'],
        },
        useBuiltIns: 'usage',
        forceAllTransforms: true,
        corejs: {
          version: 3.6,
          proposals: true,
        },
      },
    ],
  ],
});

export function createBrowserConfig({
  input,
  output,
  multiInputConfig,
  svelteConfig,
  replacements = {},
  ie11 = false as boolean,
  distElder,
  distDir,
  rootDir,
}) {
  const toReplace = {
    'process.env.componentType': "'browser'",
    'process.env.NODE_ENV': JSON.stringify(process.env.NODE_ENV),
    ...replacements,
  };

  const config = {
    cache: true,
    treeshake: production,
    input,
    output,
    plugins: [
      replace(toReplace),
      json(),
      elderSvelte({ svelteConfig, type: 'client', distElder, distDir, rootDir, legacy: ie11 }),
      nodeResolve({
        browser: true,
        dedupe: ['svelte', 'core-js'],
        preferBuiltins: true,
        rootDir: ie11 ? elderJsDir : process.cwd(),
      }),
      commonjs({ sourceMap: !production }),
    ],
  };

  // bundle splitting.
  if (multiInputConfig) {
    config.plugins.unshift(multiInputConfig);
  }

  // ie11 babel
  if (ie11) {
    config.plugins.push(babelIE11);
  }

  // if is production let's babelify everything and minify it.
  if (production) {
    // don't babel if it has been done
    if (!ie11) {
      config.plugins.push(
        babel({
          extensions: ['.js', '.mjs', '.cjs', '.html', '.svelte'],
          include: ['node_modules/**', 'src/**'],
          exclude: ['node_modules/@babel/**'],
          runtimeHelpers: true,
        }),
      );
    }

    // terser on prod
    config.plugins.push(terser());
  }

  return config;
}

export function createSSRConfig({
  input,
  output,
  svelteConfig,
  replacements = {},
  multiInputConfig,
  rootDir,
  distElder,
  distDir,
}) {
  const toReplace = {
    'process.env.componentType': "'server'",
    'process.env.NODE_ENV': JSON.stringify(process.env.NODE_ENV),
    ...replacements,
  };

  const config = {
    cache: true,
    treeshake: production,
    input,
    output,
    plugins: [
      replace(toReplace),
      json(),
      elderSvelte({ svelteConfig, type: 'ssr', distElder, distDir, rootDir }),
      nodeResolve({
        browser: false,
        dedupe: ['svelte'],
      }),
      commonjs({ sourceMap: true }),
      production && terser(),
    ],
  };
  // if we are bundle splitting include them.
  if (multiInputConfig) {
    config.plugins.unshift(multiInputConfig);
  }

  return config;
}

export default function getRollupConfig(options) {
  const defaultOptions = getDefaultRollup();
  const { svelteConfig, replacements, dev } = defaultsDeep(options, defaultOptions);
  const elderConfig = getElderConfig();
  const { $$internal, distDir, srcDir, rootDir, legacy } = elderConfig;
  const { ssrComponents, clientComponents, distElder } = $$internal;
  const relSrcDir = srcDir.replace(rootDir, '').substr(1);

  console.log(`Elder.js using rollup in ${production ? 'production' : 'development'} mode.`);

  const configs = [];

  // clear out components so there are no conflicts due to hashing.
  del.sync([`${ssrComponents}*`, `${clientComponents}*`]);
  // Add ElderJs Peer deps to public if they exist.
  [['./node_modules/intersection-observer/intersection-observer.js', './static/intersection-observer.js']].forEach(
    (dep) => {
      if (!fs.existsSync(path.resolve(rootDir, dep[0]))) {
        throw new Error(`Elder.js peer dependency not found at ${dep[0]}`);
      }
      configs.push({
        input: dep[0],
        output: [
          {
            file: path.resolve(distDir, dep[1]),
            format: 'iife',
            name: dep[1],
            plugins: [terser()],
          },
        ],
      });
    },
  );

  const { paths: pluginPaths, files: pluginFiles } = getPluginLocations(elderConfig);
  const pluginGlobs = pluginPaths.map((plugin) => `${plugin}*.svelte`);

  const components = fs.existsSync(path.resolve(srcDir, `./components/`))
    ? [
        ...new Set([
          ...glob.sync(path.resolve(srcDir, './components/*/*.svelte')),
          ...glob.sync(path.resolve(srcDir, './components/*.svelte')),
        ]),
      ]
    : [];

  configs.push(
    createSSRConfig({
      input: [
        `${relSrcDir}/layouts/*.svelte`,
        `${relSrcDir}/routes/**/*.svelte`,
        `${relSrcDir}/components/**/*.svelte`,
        ...pluginGlobs,
      ],
      output: {
        dir: ssrComponents,
        format: 'cjs',
        exports: 'auto',
        sourcemap: !production ? 'inline' : false,
      },
      multiInputConfig: multiInput({
        relative: 'src/',
        // transformOutputPath: (output) => ssrOutputPath(output),
      }),
      svelteConfig,
      replacements,
      rootDir,
      distElder,
      distDir,
    }),
  );

  configs.push(
    createBrowserConfig({
      input: [`${relSrcDir}/components/**/*.svelte`],
      output: [
        {
          dir: clientComponents,
          sourcemap: !production ? 'inline' : false,
          format: 'esm',
          entryFileNames: '[name].[hash].js',
        },
      ],
      multiInputConfig: multiInput({
        relative: 'src/',
        // transformOutputPath: (output) => `${path.basename(output)}`,
      }),
      svelteConfig,
      replacements,
      distElder,
      distDir,
      rootDir,
    }),
  );

  // legacy is only done on production or not split modes.
  if (legacy && production) {
    [...components, ...pluginFiles].forEach((cv) => {
      const file = cv.replace(`${rootDir}/`, '');
      const parsed = path.parse(cv);
      configs.push(
        createBrowserConfig({
          input: file,
          output: [
            {
              name: `___elderjs_${parsed.name}`,
              entryFileNames: '[name].[hash].js',
              dir: `${clientComponents}${path.sep}iife${path.sep}`,
              sourcemap: !production,
              format: 'iife',
            },
          ],
          svelteConfig,
          replacements,
          multiInputConfig: false,
          ie11: true,
          distElder,
          distDir,
          rootDir,
        }),
      );
    });
  }

  return configs;
}
