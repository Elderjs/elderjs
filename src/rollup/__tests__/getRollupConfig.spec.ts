/* eslint-disable global-require */
import multiInput from 'rollup-plugin-multi-input';
import path from 'path';
import { createBrowserConfig, createSSRConfig } from '../getRollupConfig';

// TODO: test replace

const fixRelativePath = (arr) => {
  arr[0].output[0].file = arr[0].output[0].file.replace(process.cwd(), '');
  return arr;
};

const common = {
  distElder: path.resolve(process.cwd(), './public/_elder/'),
  distDir: path.resolve(process.cwd(), './public/'),
  rootDir: path.resolve(process.cwd(), './'),
};

describe('#getRollupConfig', () => {
  beforeEach(() => {
    jest.resetModules();
  });

  it('createBrowserConfig works', () => {
    [true, false].forEach((sourcemap) => {
      const { plugins, ...config } = createBrowserConfig({
        input: [`./components/*/*.svelte`],
        output: {
          dir: './public/dist/svelte/',
          entryFileNames: '[name].[hash].js',
          sourcemap,
          format: 'system',
        },
        multiInputConfig: multiInput({
          // TODO: test with false
          relative: `./components`,
          transformOutputPath: (output) => `${path.basename(output)}`,
        }),
        svelteConfig: {},
        ...common,
      });
      expect(config).toEqual({
        cache: true,
        input: ['./components/*/*.svelte'],
        output: {
          dir: './public/dist/svelte/',
          entryFileNames: '[name].[hash].js',
          format: 'system',
          sourcemap,
        },
        treeshake: true,
      });
      expect(plugins).toHaveLength(8);
    });
  });

  it('createBrowserConfig multiInputConfig = false', () => {
    expect(
      createBrowserConfig({
        input: [`./components/*/*.svelte`],
        output: {
          dir: './public/dist/svelte/',
          entryFileNames: 'entry[name]-[hash].js',
          sourcemap: true,
          format: 'system',
        },
        svelteConfig: {},
        multiInputConfig: false,
        ...common,
      }).plugins.map((p) => p.name),
    ).toEqual(['replace', 'json', 'rollup-plugin-elder', 'node-resolve', 'commonjs', 'babel', 'terser']);
  });

  it('createBrowserConfig multiInputConfig = false, ie11 = true', () => {
    expect(
      createBrowserConfig({
        ...common,
        input: [`./components/*/*.svelte`],
        output: {
          dir: './public/dist/svelte/',
          entryFileNames: 'entry[name]-[hash].js',
          sourcemap: true,
          format: 'system',
        },
        svelteConfig: {},
        ie11: true,
        multiInputConfig: false,
      }).plugins.map((p) => p.name),
    ).toEqual(['replace', 'json', 'rollup-plugin-elder', 'node-resolve', 'commonjs', 'babel', 'terser']);
  });

  it('createSSRConfig works', () => {
    const { plugins, ...config } = createSSRConfig({
      ...common,
      input: [`./components/*/*.svelte`],
      output: {
        dir: './___ELDER___/compiled/',
        format: 'cjs',
        exports: 'auto',
      },
      multiInputConfig: multiInput({
        relative: `./components`,
        transformOutputPath: (output) => `${path.basename(output)}`,
      }),
      svelteConfig: {
        preprocess: [
          {
            style: ({ content }) => {
              return content.toUpperCase();
            },
          },
        ],
      },
      rootDir: 'test',
    });
    expect(config).toEqual({
      cache: true,
      input: ['./components/*/*.svelte'],
      output: {
        dir: './___ELDER___/compiled/',
        exports: 'auto',
        format: 'cjs',
      },
      treeshake: true,
    });

    expect(plugins).toHaveLength(7);
  });

  it('createSSRConfig multiInputConfig = false', () => {
    expect(
      createSSRConfig({
        ...common,
        input: [`./components/*/*.svelte`],
        output: {
          dir: './___ELDER___/compiled/',
          format: 'cjs',
          exports: 'auto',
        },
        svelteConfig: {
          preprocess: [
            {
              style: ({ content }) => {
                return content.toUpperCase();
              },
            },
          ],
        },
        rootDir: 'test',
        multiInputConfig: false,
      }).plugins.map((p) => p.name),
    ).toEqual(['replace', 'json', 'rollup-plugin-elder', 'node-resolve', 'commonjs', 'terser']);
  });

  it('getRollupConfig - throws error if intersection-observer doesnt exist', () => {
    jest.mock('../../utils/validations.ts', () => ({
      getDefaultRollup: () => ({}),
    }));

    jest.mock('del');
    // getElderConfig() mock
    jest.mock('../../utils/getConfig', () => () => ({
      $$internal: {
        clientComponents: 'test/public/svelte',
        ssrComponents: 'test/___ELDER___/compiled',
      },
      distDir: './dist',
      srcDir: './src',
      rootDir: './',
      plugins: {},
      legacy: false,
    }));
    jest.mock('fs-extra', () => ({
      existsSync: jest.fn().mockImplementationOnce(() => false),
    }));
    expect(() => require('../getRollupConfig').default()).toThrow(
      `Elder.js peer dependency not found at ./node_modules/intersection-observer/intersection-observer.js`,
    );
  });

  it('getRollupConfig as a whole works - default options', () => {
    jest.mock('../../utils/validations.ts', () => ({
      getDefaultRollup: () => ({
        replacements: {},
        dev: { splitComponents: false },
        svelteConfig: {},
      }),
    }));
    jest.mock('../../utils/getPluginLocations', () => () => ({
      paths: ['/src/plugins/elderjs-plugin-reload/'],
      files: [
        '/src/plugins/elderjs-plugin-reload/SimplePlugin.svelte',
        '/src/plugins/elderjs-plugin-reload/Test.svelte',
      ],
    }));
    // getElderConfig() mock
    jest.mock('../../utils/getConfig', () => () => ({
      $$internal: {
        clientComponents: 'test/public/svelte',
        ssrComponents: 'test/___ELDER___/compiled',
      },
      distDir: './dist',
      srcDir: './src',
      rootDir: './',
      plugins: {
        pluginA: {},
        pluginB: {},
      },
      legacy: false,
    }));

    jest.mock('del');
    jest.mock('fs-extra', () => ({
      existsSync: jest.fn().mockImplementation(() => true),
    }));

    const svelteConfig = {
      preprocess: [
        {
          style: ({ content }) => {
            return content.toUpperCase();
          },
        },
      ],
    };

    // would be nice to mock getPluginPaths if it's extracted to separate file
    const configs = fixRelativePath(require('../getRollupConfig').default({ svelteConfig }));
    expect(configs).toHaveLength(3);
    expect(configs).toMatchSnapshot();
  });

  it('getRollupConfig as a whole works - legacy: true', () => {
    jest.mock('../../utils/validations.ts', () => ({
      getDefaultRollup: () => ({
        replacements: {},
        dev: { splitComponents: false },
        svelteConfig: {},
      }),
    }));
    jest.mock('../../utils/getPluginLocations', () => () => ({
      paths: ['/src/plugins/elderjs-plugin-reload/'],
      files: [
        '/src/plugins/elderjs-plugin-reload/SimplePlugin.svelte',
        '/src/plugins/elderjs-plugin-reload/Test.svelte',
      ],
    }));
    // getElderConfig() mock
    jest.mock('../../utils/getConfig', () => () => ({
      $$internal: {
        clientComponents: 'test/public/svelte',
        ssrComponents: 'test/___ELDER___/compiled',
      },
      distDir: './dist',
      srcDir: './src',
      rootDir: './',
      plugins: {
        pluginA: {},
        pluginB: {},
      },
      legacy: true,
    }));

    jest.mock('fs-extra', () => ({
      existsSync: jest.fn().mockImplementation(() => true),
    }));
    jest.mock('glob', () => ({
      sync: jest
        .fn()
        .mockImplementationOnce(() => ['pluginAComponent1.svelte', 'pluginAComponent2.svelte', 'AnythingCanBeHere']) // getPluginPaths
        .mockImplementationOnce(() => ['pluginBComponent1.svelte', 'pluginBComponent2.svelte', 'AnythingCanBeHere']) // getPluginPaths
        .mockImplementationOnce(() => ['NestedSrcComponent.svelte']) // nested
        .mockImplementationOnce(() => ['SrcComponent.svlete', 'SrcComponent2.svelte']) // src
        .mockImplementationOnce(() => ['LegacyPluginAComponent3.svelte', 'LegacyPluginAComponent4.svelte']) // legacy pluginA components
        .mockImplementationOnce(() => ['LegacyPluginBComponent3.svelte']), // legacy pluginB components
    }));

    const svelteConfig = {
      preprocess: [
        {
          style: ({ content }) => {
            return content.toUpperCase();
          },
        },
      ],
    };

    // would be nice to mock getPluginPaths if it's extracted to separate file
    const configs = fixRelativePath(require('../getRollupConfig').default({ svelteConfig }));
    expect(configs).toHaveLength(10);
    expect(configs).toMatchSnapshot();
  });
});
