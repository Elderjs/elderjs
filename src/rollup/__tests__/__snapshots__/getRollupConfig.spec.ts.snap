// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#getRollupConfig getRollupConfig as a whole works - default options 1`] = `
Array [
  Object {
    "input": "./node_modules/intersection-observer/intersection-observer.js",
    "output": Array [
      Object {
        "file": "/dist/static/intersection-observer.js",
        "format": "iife",
        "name": "./static/intersection-observer.js",
        "plugins": Array [
          Object {
            "name": "terser",
            "renderChunk": [Function],
          },
        ],
      },
    ],
  },
  Object {
    "cache": true,
    "input": Array [
      "rc/layouts/*.svelte",
      "rc/routes/**/*.svelte",
      "rc/components/**/*.svelte",
      "/src/plugins/elderjs-plugin-reload/*.svelte",
    ],
    "output": Object {
      "dir": "test/___ELDER___/compiled",
      "exports": "auto",
      "format": "cjs",
      "sourcemap": false,
    },
    "plugins": Array [
      Object {
        "options": [Function],
        "pluginName": "rollup-plugin-multi-input",
      },
      Object {
        "name": "replace",
        "renderChunk": [Function],
        "transform": [Function],
      },
      Object {
        "name": "json",
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "load": [Function],
        "name": "rollup-plugin-elder",
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "getPackageInfoForId": [Function],
        "load": [Function],
        "name": "node-resolve",
        "resolveId": [Function],
      },
      Object {
        "buildStart": [Function],
        "load": [Function],
        "name": "commonjs",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "name": "terser",
        "renderChunk": [Function],
      },
    ],
    "treeshake": true,
  },
  Object {
    "cache": true,
    "input": Array [
      "rc/components/**/*.svelte",
    ],
    "output": Array [
      Object {
        "dir": "test/public/svelte",
        "entryFileNames": "[name].[hash].js",
        "format": "esm",
        "sourcemap": false,
      },
    ],
    "plugins": Array [
      Object {
        "options": [Function],
        "pluginName": "rollup-plugin-multi-input",
      },
      Object {
        "name": "replace",
        "renderChunk": [Function],
        "transform": [Function],
      },
      Object {
        "name": "json",
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "load": [Function],
        "name": "rollup-plugin-elder",
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "getPackageInfoForId": [Function],
        "load": [Function],
        "name": "node-resolve",
        "resolveId": [Function],
      },
      Object {
        "buildStart": [Function],
        "load": [Function],
        "name": "commonjs",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "load": [Function],
        "name": "babel",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "name": "terser",
        "renderChunk": [Function],
      },
    ],
    "treeshake": true,
  },
]
`;

exports[`#getRollupConfig getRollupConfig as a whole works - legacy: true 1`] = `
Array [
  Object {
    "input": "./node_modules/intersection-observer/intersection-observer.js",
    "output": Array [
      Object {
        "file": "/dist/static/intersection-observer.js",
        "format": "iife",
        "name": "./static/intersection-observer.js",
        "plugins": Array [
          Object {
            "name": "terser",
            "renderChunk": [Function],
          },
        ],
      },
    ],
  },
  Object {
    "cache": true,
    "input": Array [
      "rc/layouts/*.svelte",
      "rc/routes/**/*.svelte",
      "rc/components/**/*.svelte",
      "/src/plugins/elderjs-plugin-reload/*.svelte",
    ],
    "output": Object {
      "dir": "test/___ELDER___/compiled",
      "exports": "auto",
      "format": "cjs",
      "sourcemap": false,
    },
    "plugins": Array [
      Object {
        "options": [Function],
        "pluginName": "rollup-plugin-multi-input",
      },
      Object {
        "name": "replace",
        "renderChunk": [Function],
        "transform": [Function],
      },
      Object {
        "name": "json",
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "load": [Function],
        "name": "rollup-plugin-elder",
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "getPackageInfoForId": [Function],
        "load": [Function],
        "name": "node-resolve",
        "resolveId": [Function],
      },
      Object {
        "buildStart": [Function],
        "load": [Function],
        "name": "commonjs",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "name": "terser",
        "renderChunk": [Function],
      },
    ],
    "treeshake": true,
  },
  Object {
    "cache": true,
    "input": Array [
      "rc/components/**/*.svelte",
    ],
    "output": Array [
      Object {
        "dir": "test/public/svelte",
        "entryFileNames": "[name].[hash].js",
        "format": "esm",
        "sourcemap": false,
      },
    ],
    "plugins": Array [
      Object {
        "options": [Function],
        "pluginName": "rollup-plugin-multi-input",
      },
      Object {
        "name": "replace",
        "renderChunk": [Function],
        "transform": [Function],
      },
      Object {
        "name": "json",
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "load": [Function],
        "name": "rollup-plugin-elder",
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "getPackageInfoForId": [Function],
        "load": [Function],
        "name": "node-resolve",
        "resolveId": [Function],
      },
      Object {
        "buildStart": [Function],
        "load": [Function],
        "name": "commonjs",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "load": [Function],
        "name": "babel",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "name": "terser",
        "renderChunk": [Function],
      },
    ],
    "treeshake": true,
  },
  Object {
    "cache": true,
    "input": "pluginAComponent1.svelte",
    "output": Array [
      Object {
        "dir": "test/public/svelte/iife/",
        "entryFileNames": "[name].[hash].js",
        "format": "iife",
        "name": "___elderjs_pluginAComponent1",
        "sourcemap": false,
      },
    ],
    "plugins": Array [
      Object {
        "name": "replace",
        "renderChunk": [Function],
        "transform": [Function],
      },
      Object {
        "name": "json",
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "load": [Function],
        "name": "rollup-plugin-elder",
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "getPackageInfoForId": [Function],
        "load": [Function],
        "name": "node-resolve",
        "resolveId": [Function],
      },
      Object {
        "buildStart": [Function],
        "load": [Function],
        "name": "commonjs",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "load": [Function],
        "name": "babel",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "name": "terser",
        "renderChunk": [Function],
      },
    ],
    "treeshake": true,
  },
  Object {
    "cache": true,
    "input": "pluginAComponent2.svelte",
    "output": Array [
      Object {
        "dir": "test/public/svelte/iife/",
        "entryFileNames": "[name].[hash].js",
        "format": "iife",
        "name": "___elderjs_pluginAComponent2",
        "sourcemap": false,
      },
    ],
    "plugins": Array [
      Object {
        "name": "replace",
        "renderChunk": [Function],
        "transform": [Function],
      },
      Object {
        "name": "json",
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "load": [Function],
        "name": "rollup-plugin-elder",
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "getPackageInfoForId": [Function],
        "load": [Function],
        "name": "node-resolve",
        "resolveId": [Function],
      },
      Object {
        "buildStart": [Function],
        "load": [Function],
        "name": "commonjs",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "load": [Function],
        "name": "babel",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "name": "terser",
        "renderChunk": [Function],
      },
    ],
    "treeshake": true,
  },
  Object {
    "cache": true,
    "input": "AnythingCanBeHere",
    "output": Array [
      Object {
        "dir": "test/public/svelte/iife/",
        "entryFileNames": "[name].[hash].js",
        "format": "iife",
        "name": "___elderjs_AnythingCanBeHere",
        "sourcemap": false,
      },
    ],
    "plugins": Array [
      Object {
        "name": "replace",
        "renderChunk": [Function],
        "transform": [Function],
      },
      Object {
        "name": "json",
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "load": [Function],
        "name": "rollup-plugin-elder",
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "getPackageInfoForId": [Function],
        "load": [Function],
        "name": "node-resolve",
        "resolveId": [Function],
      },
      Object {
        "buildStart": [Function],
        "load": [Function],
        "name": "commonjs",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "load": [Function],
        "name": "babel",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "name": "terser",
        "renderChunk": [Function],
      },
    ],
    "treeshake": true,
  },
  Object {
    "cache": true,
    "input": "pluginBComponent1.svelte",
    "output": Array [
      Object {
        "dir": "test/public/svelte/iife/",
        "entryFileNames": "[name].[hash].js",
        "format": "iife",
        "name": "___elderjs_pluginBComponent1",
        "sourcemap": false,
      },
    ],
    "plugins": Array [
      Object {
        "name": "replace",
        "renderChunk": [Function],
        "transform": [Function],
      },
      Object {
        "name": "json",
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "load": [Function],
        "name": "rollup-plugin-elder",
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "getPackageInfoForId": [Function],
        "load": [Function],
        "name": "node-resolve",
        "resolveId": [Function],
      },
      Object {
        "buildStart": [Function],
        "load": [Function],
        "name": "commonjs",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "load": [Function],
        "name": "babel",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "name": "terser",
        "renderChunk": [Function],
      },
    ],
    "treeshake": true,
  },
  Object {
    "cache": true,
    "input": "pluginBComponent2.svelte",
    "output": Array [
      Object {
        "dir": "test/public/svelte/iife/",
        "entryFileNames": "[name].[hash].js",
        "format": "iife",
        "name": "___elderjs_pluginBComponent2",
        "sourcemap": false,
      },
    ],
    "plugins": Array [
      Object {
        "name": "replace",
        "renderChunk": [Function],
        "transform": [Function],
      },
      Object {
        "name": "json",
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "load": [Function],
        "name": "rollup-plugin-elder",
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "getPackageInfoForId": [Function],
        "load": [Function],
        "name": "node-resolve",
        "resolveId": [Function],
      },
      Object {
        "buildStart": [Function],
        "load": [Function],
        "name": "commonjs",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "load": [Function],
        "name": "babel",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "name": "terser",
        "renderChunk": [Function],
      },
    ],
    "treeshake": true,
  },
  Object {
    "cache": true,
    "input": "/src/plugins/elderjs-plugin-reload/SimplePlugin.svelte",
    "output": Array [
      Object {
        "dir": "test/public/svelte/iife/",
        "entryFileNames": "[name].[hash].js",
        "format": "iife",
        "name": "___elderjs_SimplePlugin",
        "sourcemap": false,
      },
    ],
    "plugins": Array [
      Object {
        "name": "replace",
        "renderChunk": [Function],
        "transform": [Function],
      },
      Object {
        "name": "json",
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "load": [Function],
        "name": "rollup-plugin-elder",
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "getPackageInfoForId": [Function],
        "load": [Function],
        "name": "node-resolve",
        "resolveId": [Function],
      },
      Object {
        "buildStart": [Function],
        "load": [Function],
        "name": "commonjs",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "load": [Function],
        "name": "babel",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "name": "terser",
        "renderChunk": [Function],
      },
    ],
    "treeshake": true,
  },
  Object {
    "cache": true,
    "input": "/src/plugins/elderjs-plugin-reload/Test.svelte",
    "output": Array [
      Object {
        "dir": "test/public/svelte/iife/",
        "entryFileNames": "[name].[hash].js",
        "format": "iife",
        "name": "___elderjs_Test",
        "sourcemap": false,
      },
    ],
    "plugins": Array [
      Object {
        "name": "replace",
        "renderChunk": [Function],
        "transform": [Function],
      },
      Object {
        "name": "json",
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "load": [Function],
        "name": "rollup-plugin-elder",
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "getPackageInfoForId": [Function],
        "load": [Function],
        "name": "node-resolve",
        "resolveId": [Function],
      },
      Object {
        "buildStart": [Function],
        "load": [Function],
        "name": "commonjs",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "load": [Function],
        "name": "babel",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "name": "terser",
        "renderChunk": [Function],
      },
    ],
    "treeshake": true,
  },
]
`;
